### DISCUSSION TITLE
Odd behavior with the jwt-token-validator

### DISCUSSION THREAD

**Discussion by phil-hotchkiss** [mia-platform/community#496] [Q&A] [answered]
### What product are you having troubles with?

Console

### What Console version are you using?

12.4.0

### Description

We are seeing some odd behavior that is eerily similar to the situation described in [this](https://github.com/mia-platform/community/discussions/386) thread. 

We have a call stack that looks like this: 

client application -> azure api management -> envoy -> authorization-service -> jwt-token-validator -> Mia microservice. 

The error seems to occur only with newly provisioned tokens and only right after they are provisioned. When the token is subsequently re-used, it seems to work as expected. I don't know why this is an issue, but it seems to behaving like a clock skew problem. 

The Azure API Management layer validates the client application's jwt successfully, however, the jwt-token-validator returns the following log lines: 

```
2024-05-15 10:00:32.778	{"host":{"hostname":"jwt-token-validator"},"http":{"request":{"method":"GET","userAgent":{"original":"Go-http-client/1.1"}},"response":{"statusCode":400,"body":{"bytes":29}}},"level":30,"msg":"request completed","reqId":"215b86a5-a2fb-4103-b2ee-eaf63d51b521","responseTime":0,"time":1715785232778,"url":{"path":"/verify"}}

2024-05-15 10:00:32.778	{"error":{"message":"token found in cache jwks"},"level":50,"msg":"JWT is not valid.","reqId":"215b86a5-a2fb-4103-b2ee-eaf63d51b521","time":1715785232778}

2024-05-15 10:00:32.778	{"host":{"hostname":"jwt-token-validator"},"http":{"request":{"method":"GET","userAgent":{"original":"Go-http-client/1.1"}}},"level":10,"msg":"incoming request","reqId":"215b86a5-a2fb-4103-b2ee-eaf63d51b521","time":1715785232778,"url":{"path":"/verify"}}

```

The log message doesn't make a lot of sense to me. The JWKS endpoint we've provided doesn't have a rotating key (it only rotates a couple of times per year). Again, we've validated the token prior to hitting the jwt-token-validator, so it doesn't make a lot of sense that it (the token) doesn't work when it is first used and then subsequently succeeds. 

Our environment variables are as follows: 

```
SERVICE_VERSION: 2.0.2
USERINFO_URL: http://jwt-token-validator/verify
CUSTOM_USER_ID_KEY: sub
HEADERS_TO_PROXY: authorization, client-type, host, x-forwarded-host
USER_PROPERTIES_TO_PROXY: sub
CLIENT_TYPE_HEADER_KEY: client-type
BACKOFFICE_USERINFOR_URL: http://jwt-token-validator/verify
CUSTOM_PERMISSIONS_KEY: permissions
AUTHORIZATION_HEADERS_TO_PROXY: authorization
```

#### Microservice Versions
- jwt-token-validator:1.1.1
- Envoy: envoyproxy/envoy:v1.30-latest
- authorization-service:2.4.2

### Actual Outcome

Occasional 401s with error notes provided above when token is newly minted. Inconsistently occurring. 

### Expected Outcome

Token is valid, jwt-token-validator/verify should be returning `200 OK`.

**Comment by Danielecina** [comment]
Hi @phil-hotchkiss. This issue reminds me a lot of this one https://github.com/mia-platform/community/discussions/386. 
A colleague solved the problem at the time. I tagged him with the idea that he can help you specifically. 
@Lucascanna can you think of how you resolved at that time?

**Comment by lucascanna** [comment, marked as answer]
Hello Phil,

I can confirm that, as you mentioned, this is a clock skew problem, i.e. the clock of the machine where the token is issued and the clock of the machine where the token is verified are out of sync.
The "jwt-token-validator" service uses the [golang-jwt](https://github.com/golang-jwt/jwt) **v4** library to validate tokens. Unfortunately, the v4 of the library does not support a clock tolerance option.

Specifically, in your case the problem is not with the _nbf_ claim but the _iat_ claim. Indeed, the default behaviour of the v4 library (that has been changed in v5) is to also check that the _iat_ claim is not in the future.
We will plan the migration of the "jwt-token-validator" to [golang-jwt](https://github.com/golang-jwt/jwt) **v5** and let you know once done.

In the meanwhile, one quick and dirty solution could be to avoid the _iat_ claim in the token. According to [the docs](https://pkg.go.dev/github.com/golang-jwt/jwt/v4@v4.5.0#MapClaims.Valid), this should also fix the problem.

Best Regards.

**Comment by galatifederico** [comment]
Hello @phil-hotchkiss,
The version **1.1.2** of the plugin fixed the issue. 

Best Regards.
